### ISSUE TITLE
Type narrowing error in async function while return type mark as `NoReturn`

### ISSUE THREAD

**Issue by uf-fipper** [microsoft/pyright#5876]
## Environment data

-   Language Server version: 2023.8.41 (23 August 2023) PreRelease
-   OS and version: win11
-   Python version (& distribution if applicable, e.g. Anaconda): 3.9

## Code Snippet

```python
import asyncio
from typing import NoReturn, Optional

async def raise_exc() -> NoReturn:
    raise Exception()

def get() -> Optional[int]:
    ...

async def main():
    x = get()  # int | None
    if x is None:
        await raise_exc()
    x.as_integer_ratio()  # pylance: "as_integer_ratio" is not a known member of "None"

asyncio.run(main())
```

## Expected behavior

after `await raise_exc()`, x should be inferred as an integer.

## Actual behavior

it is inferred as `int | None`.

## Extra

but in the synchronous function, pylance will provide the correct inference:

```python
from typing import NoReturn, Optional

def raise_exc() -> NoReturn:
    raise Exception()

def get() -> Optional[int]:
    ...

def main():
    x = get()  # int | None
    if x is None:
        raise_exc()
    x.as_integer_ratio()  # int

main()
```

**Comment by erictraut** [comment]
This was a regression related to the fix for https://github.com/microsoft/pyright/issues/5639.

It will be fixed in the next release of pyright.

**Comment by erictraut** [comment]
This is addressed in pyright 1.1.326, which I just published. It will also be included in a future release of pylance.
